scanpipe/pipes/purldb.py

```diff
@@ -168,11 +174,152 @@ def match_directory(fingerprint, timeout=DEFAULT_TIMEOUT, api_url=PURLDB_API_URL
+    unsupported_vers_count = 0
+
+    progress = LoopProgress(batches_count, logger)
+
+    for batch in progress.iter(package_batches):
+        if response := submit_purls(packages=batch):
+            queued_packages_count += response.get("queued_packages_count", 0)
+            unqueued_packages_count += response.get("unqueued_packages_count", 0)
+            unsupported_packages_count += response.get("unsupported_packages_count", 0)
+            unsupported_vers_count += response.get("unsupported_vers_count", 0)
+
+    if queued_packages_count > 0:
+        logger(
+            f"Successfully queued {queued_packages_count:,d} "
+            f"PURLs for indexing in PurlDB"
+        )
+
+    if unqueued_packages_count > 0:
+        logger(
+            f"{unqueued_packages_count:,d} PURLs were already "
+            f"present in PurlDB index queue"
+        )
+
+    if unsupported_packages_count > 0:
+        logger(f"Couldn't index {unsupported_packages_count:,d} unsupported PURLs")
+
+    if unsupported_vers_count > 0:
+        logger(f"Couldn't index {unsupported_vers_count:,d} unsupported vers")
+
+
+def get_unique_resolved_purls(project):
+    """Return PURLs from project's resolved DiscoveredDependencies."""
+    packages_resolved = project.discovereddependencies.filter(is_resolved=True)
+
+    distinct_results = packages_resolved.values("type", "namespace", "name", "version")
+
+    distinct_combinations = {tuple(item.values()) for item in distinct_results}
+    return {str(PackageURL(*values)) for values in distinct_combinations}
+
+
+def get_unique_unresolved_purls(project):
+    """Return PURLs from project's unresolved DiscoveredDependencies."""
+    packages_unresolved = project.discovereddependencies.filter(
+        is_resolved=False
+    ).exclude(extracted_requirement="*")
+
+    distinct_unresolved_results = packages_unresolved.values(
+        "type", "namespace", "name", "extracted_requirement"
+    )
+
+    distinct_unresolved = {tuple(item.values()) for item in distinct_unresolved_results}
+
+    packages = set()
+    for item in distinct_unresolved:
+        pkg_type, namespace, name, extracted_requirement = item
+        if range_class := RANGE_CLASS_BY_SCHEMES.get(pkg_type):
+            try:
+                vers = range_class.from_native(extracted_requirement)
+            except InvalidVersionRange:
+                continue
+
+            if not vers.constraints:
+                continue
+
+            purl = PackageURL(type=pkg_type, namespace=namespace, name=name)
+            packages.add((str(purl), str(vers)))
+
+    return packages
+
+
+def populate_purldb_with_discovered_packages(project, logger=None):
+    """Add DiscoveredPackage to PurlDB."""
+    discoveredpackages = project.discoveredpackages.all()
+    packages = [{"purl": pkg.purl} for pkg in discoveredpackages]
+
+    feed_purldb(
+        packages=packages,
+        msg=f"Populating PurlDB with {len(packages):,d} PURLs from DiscoveredPackage",
```

Comment:
Any reasons to delegate the msg logging to the `feed_purldb` function instead of doing it before the function call?

Reply:
Saving some code duplication ;)
We'll end up duplicating this same piece of code in four different places.
```python
    if logger:
        if not packages:
            logger("No PURLs found. Skipping.")
            return
        logger("Some logging msg")
```

Reply:
@keshav-space First, this logic is broken: if the logger is not set, the `return` is not reached when `not packages`.

Being DRY is important for code that actually does something, not so much for boilerplate code such as logging.
Here, I don't see the the benefit of delegating the logging to the feed_purldb as a good trade-off, I think it actually makes the code more complex.

What about this:
- Instead of `logger=None`, let's use the default module logger `logger=logger.info`, this will remove the need to check for the logger availability and also log to the console by default.
- Keep the main logging message before the `feed_purldb` call (even if the count is 0)
- Have the simplified following in the feed_pruldb:
```
if not packages:
    logger("No PURLs found. Skipping.")
    return
```

Reply:
Thanks @tdruez , I have simplified the logging as per your suggestion.